### Assignment -- Recommendation System

Problem statement.

Build a recommender system by using cosine simillarties score.

Dataset --> bookRS.csv

In [1]:
#Importing the Required Libraries
import pandas as pd
import numpy as np
#Calculating Cosine Similarity between Users
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine

In [2]:
#Loading the Dataset
books_df = pd.read_csv('C:/Users/Akaash/Downloads/bookRS.csv')
books_df.head()

Unnamed: 0  User.ID                                         Book.Title  \
0           1   276726                                Classical Mythology   
1           2   276729                                       Clara Callan   
2           3   276729                               Decision in Normandy   
3           4   276736  Flu: The Story of the Great Influenza Pandemic...   
4           5   276737                             The Mummies of Urumchi   

   Book.Rating  
0            5  
1            3  
2            6  
3            8  
4            6

Inference: Droping the unnamed columns as it is just index column

In [3]:
#Dropping the Unnamed Column
books_df = books_df.iloc[:,1:]
books_df.head()

User.ID                                         Book.Title  Book.Rating
0   276726                                Classical Mythology            5
1   276729                                       Clara Callan            3
2   276729                               Decision in Normandy            6
3   276736  Flu: The Story of the Great Influenza Pandemic...            8
4   276737                             The Mummies of Urumchi            6

In [4]:
#Checking the Shape of the Dataset
books_df.shape

(10000, 3)

In [5]:
#number of unique users in the dataset
len(books_df['User.ID'].unique())

2182

In [6]:
#number of unique Books in the dataset
len(books_df['Book.Title'].unique())

9659

Inference:

Total/Unqiue Users = 2182, Total/Unqiue Books = 9659

Here Product is Books and users are customer, so all the unique user should come in row side, and unique Books on the column side. So now we can use crosstab / Pivot table

After Using a Crosstab / Pivot Table we will get (2182,9659)--(rows,columns)

In [7]:
#Creating a Pivot Table DataFrame
user_books_df = books_df.pivot_table(index='User.ID', columns='Book.Title', values='Book.Rating').reset_index(drop=True)
user_books_df

Book.Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
0                            NaN                                         NaN   
1                            NaN                                         NaN   
2                            NaN                                         NaN   
3                            NaN                                         NaN   
4                            NaN                                         NaN   
...                          ...                                         ...   
2177                         NaN                                         NaN   
2178                         NaN                                         NaN   
2179                         NaN                                         NaN   
2180                         NaN                                         NaN   
2181                         NaN                                         NaN   

Book.Title   Repairing PC Drives &amp  '48  \
0                                 NaN  NaN   
1                                 NaN  NaN   
2                                 NaN  NaN   
3                                 NaN  NaN   
4                                 NaN  NaN   
...                               ...  ...   
2177                              NaN  NaN   
2178                              NaN  NaN   
2179                              NaN  NaN   
2180                              NaN  NaN   
2181                              NaN  NaN   

Book.Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
0                                                         NaN                       
1                                                         NaN                       
2                                                         NaN                       
3                                                         NaN                       
4                                                         NaN                       
...                                                       ...                       
2177                                                      NaN                       
2178                                                      NaN                       
2179                                                      NaN                       
2180                                                      NaN                       
2181                                                      NaN                       

Book.Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
0                                                         NaN              
1                                                         NaN              
2                                                         NaN              
3                                                         NaN              
4                                                         NaN              
...                                                       ...              
2177                                                      NaN              
2178                                                      NaN              
2179                                                      NaN              
2180                                                      NaN              
2181                                                      NaN              

Book.Title  01-01-00: A Novel of the Millennium  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   
...                                         ...   
2177                                        NaN   
2178                                        NaN   
2179                                        NaN   
2180                                      

Inference: Inference: Now you can see we have (2182,9659) shape with proper index,as the previous index was reset using -->reset_index(drop=True)

Where drop = True drops original index

In [8]:
#Replaceing index with userid
user_books_df.index = books_df['User.ID'].unique()
user_books_df

Book.Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
276726                       NaN                                         NaN   
276729                       NaN                                         NaN   
276736                       NaN                                         NaN   
276737                       NaN                                         NaN   
276744                       NaN                                         NaN   
...                          ...                                         ...   
162107                       NaN                                         NaN   
162109                       NaN                                         NaN   
162113                       NaN                                         NaN   
162121                       NaN                                         NaN   
162129                       NaN                                         NaN   

Book.Title   Repairing PC Drives &amp  '48  \
276726                            NaN  NaN   
276729                            NaN  NaN   
276736                            NaN  NaN   
276737                            NaN  NaN   
276744                            NaN  NaN   
...                               ...  ...   
162107                            NaN  NaN   
162109                            NaN  NaN   
162113                            NaN  NaN   
162121                            NaN  NaN   
162129                            NaN  NaN   

Book.Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
276726                                                    NaN                       
276729                                                    NaN                       
276736                                                    NaN                       
276737                                                    NaN                       
276744                                                    NaN                       
...                                                       ...                       
162107                                                    NaN                       
162109                                                    NaN                       
162113                                                    NaN                       
162121                                                    NaN                       
162129                                                    NaN                       

Book.Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
276726                                                    NaN              
276729                                                    NaN              
276736                                                    NaN              
276737                                                    NaN              
276744                                                    NaN              
...                                                       ...              
162107                                                    NaN              
162109                                                    NaN              
162113                                                    NaN              
162121                                                    NaN              
162129                                                    NaN              

Book.Title  01-01-00: A Novel of the Millennium  \
276726                                      NaN   
276729                                      NaN   
276736                                      NaN   
276737                                      NaN   
276744                                      NaN   
...                                         ...   
162107                                      NaN   
162109                                      NaN   
162113                                      NaN   
162121                                    

Inference: Replacing the index column with user's user id

In [9]:
#Impute those NaNs with 0 values
user_books_df.fillna(0, inplace=True)
user_books_df

Book.Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
276726                       0.0                                         0.0   
276729                       0.0                                         0.0   
276736                       0.0                                         0.0   
276737                       0.0                                         0.0   
276744                       0.0                                         0.0   
...                          ...                                         ...   
162107                       0.0                                         0.0   
162109                       0.0                                         0.0   
162113                       0.0                                         0.0   
162121                       0.0                                         0.0   
162129                       0.0                                         0.0   

Book.Title   Repairing PC Drives &amp  '48  \
276726                            0.0  0.0   
276729                            0.0  0.0   
276736                            0.0  0.0   
276737                            0.0  0.0   
276744                            0.0  0.0   
...                               ...  ...   
162107                            0.0  0.0   
162109                            0.0  0.0   
162113                            0.0  0.0   
162121                            0.0  0.0   
162129                            0.0  0.0   

Book.Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
276726                                                    0.0                       
276729                                                    0.0                       
276736                                                    0.0                       
276737                                                    0.0                       
276744                                                    0.0                       
...                                                       ...                       
162107                                                    0.0                       
162109                                                    0.0                       
162113                                                    0.0                       
162121                                                    0.0                       
162129                                                    0.0                       

Book.Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
276726                                                    0.0              
276729                                                    0.0              
276736                                                    0.0              
276737                                                    0.0              
276744                                                    0.0              
...                                                       ...              
162107                                                    0.0              
162109                                                    0.0              
162113                                                    0.0              
162121                                                    0.0              
162129                                                    0.0              

Book.Title  01-01-00: A Novel of the Millennium  \
276726                                      0.0   
276729                                      0.0   
276736                                      0.0   
276737                                      0.0   
276744                                      0.0   
...                                         ...   
162107                                      0.0   
162109                                      0.0   
162113                                      0.0   
162121                                    

Inference: Filling NA Values with 0 as considering that the users haven't given rating to those books which they didn't Read.

#### Checking Similiarity --Cosine Based Similarity

In [10]:
#Checking Similiarity by using Cosine based Similarity -- passing an array of values
user_sim = 1 - pairwise_distances( user_books_df.values,metric='cosine')
user_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [11]:
#Store the results of Cosine-Base Similarity in a dataframe
user_sim_df = pd.DataFrame(user_sim)

#Set the index and column names to user ids 
user_sim_df.index = books_df['User.ID'].unique()
user_sim_df.columns = books_df['User.ID'].unique()
#Getting First 5 Row and columns
user_sim_df.iloc[0:5,0:5]

276726  276729  276736  276737  276744
276726     1.0     0.0     0.0     0.0     0.0
276729     0.0     1.0     0.0     0.0     0.0
276736     0.0     0.0     1.0     0.0     0.0
276737     0.0     0.0     0.0     1.0     0.0
276744     0.0     0.0     0.0     0.0     1.0

Inference: Created the matrix of Similiarity, 1 = Similiarity between Two Users, 0 = No Similiarity
    
Now as you can see similiarity between ourselves will be always 1 so droping them i.e diagonal element

In [12]:
#Filling Diagonal to 0
np.fill_diagonal(user_sim, 0)
#Getting First 5 Row and columns
user_sim_df.iloc[0:5, 0:5]

276726  276729  276736  276737  276744
276726     0.0     0.0     0.0     0.0     0.0
276729     0.0     0.0     0.0     0.0     0.0
276736     0.0     0.0     0.0     0.0     0.0
276737     0.0     0.0     0.0     0.0     0.0
276744     0.0     0.0     0.0     0.0     0.0

In [13]:
#Most Similar Users
user_sim_df.idxmax(axis=1)[0:]

276726    276726
276729    276726
276736    276726
276737    276726
276744    276726
           ...  
162107    276726
162109    276726
162113    161453
162121    276726
162129    276726
Length: 2182, dtype: int64

Inference: All these users have simliarity between them.

In [14]:
#Checking Details of 162113 & 161453 Users to Recommend
books_df[(books_df['User.ID']==162113) | (books_df['User.ID']==161453)]

User.ID                                         Book.Title  Book.Rating
8959   161453  Bread, Tomato, Garlic: Quick Cooking With 3 Ma...            9
8960   161453  The Ubiquitous Shrimp: From Simple to Exotic, ...            8
9989   162113       The Cape Ann (Contemporary American Fiction)            8

In [15]:
#Creating DataFrame of Two User_id for Join Statement
user_1=books_df[(books_df['User.ID']==162113)]
user_2=books_df[(books_df['User.ID']==161453)]
print(user_1['Book.Title'])
print(user_2['Book.Title'])

9989    The Cape Ann (Contemporary American Fiction)
Name: Book.Title, dtype: object
8959    Bread, Tomato, Garlic: Quick Cooking With 3 Ma...
8960    The Ubiquitous Shrimp: From Simple to Exotic, ...
Name: Book.Title, dtype: object


In [16]:
#Join Books Reads by both and then we can recommend the ones which have not been Read-- Joins like SQL
pd.merge(user_1,user_2,on='Book.Title',how='outer')

User.ID_x                                         Book.Title  \
0   162113.0       The Cape Ann (Contemporary American Fiction)   
1        NaN  Bread, Tomato, Garlic: Quick Cooking With 3 Ma...   
2        NaN  The Ubiquitous Shrimp: From Simple to Exotic, ...   

   Book.Rating_x  User.ID_y  Book.Rating_y  
0            8.0        NaN            NaN  
1            NaN   161453.0            9.0  
2            NaN   161453.0            8.0

##### Inference: So now we can recommend The Book "The Cape Ann" to User.ID 161453, also recommend The Book "Bread, Tomato, Garlic"  & "The Ubiquitous Shrimp" to User.ID 162113  based on above matrix, According to Their Similarity of Rating books